[KED-1876] Allow GBQTableDataSet to optionally accept a sql query to load data

ajb7 opened the pull request:
## Description

Modify current implementation of `GBQTableDataSet` to accept custom queries to be executed in GBQ to load custom results.

Addresses:  https://github.com/quantumblacklabs/kedro/issues/442

## Development notes
Modified the `_load()` method in `GBQTableDataSet` to accept custom queries passed via `load_args` in `catalog.yml`. Eventually passing the arguments in `pandas.read_gbq()` 

## Checklist

- [ X] Read the [contributing](https://github.com/quantumblacklabs/kedro/blob/master/CONTRIBUTING.md) guidelines
- [ X] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ X] Updated the documentation to reflect the code changes
- [ X] Added a description of this change and added my name to the list of supporting contributions in the [`RELEASE.md`](https://github.com/quantumblacklabs/kedro/blob/master/RELEASE.md) file
- [ X] Added tests to cover my changes

## Notice

- [X ] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

mzjp2 commented on kedro/extras/datasets/pandas/gbq_dataset.py:
```diff
@@ -137,12 +137,18 @@ def _describe(self) -> Dict[str, Any]:
+        if "query" not in self._load_args:
+            self._load_args.update({"query": sql})
+
+        self._load_args["query"] = str(self._load_args["query"])
```
why do we need to stringify it?

mzjp2 commented on kedro/extras/datasets/pandas/gbq_dataset.py:
```diff
@@ -137,12 +137,18 @@ def _describe(self) -> Dict[str, Any]:
-            credentials=self._credentials,
-            **self._load_args
-        )
+        if "query" not in self._load_args:
+            self._load_args.update({"query": sql})
```
I think you can shorten this/make it cleaner/more readable by just doing `self._load_args.setdefault("query", sql)` :)

https://docs.python.org/3/library/stdtypes.html#dict.setdefault

mzjp2 commented on kedro/extras/datasets/pandas/gbq_dataset.py:
```diff
@@ -137,12 +137,18 @@ def _describe(self) -> Dict[str, Any]:
+                credentials=self._credentials,
+                **self._load_args
+            )
+        except ValueError:
+            raise ValueError()
```
why is this needed? can you not skip the `try/except`? We don't do this for any of our other datasets afaics

mzjp2 commented on tests/extras/datasets/pandas/test_gbq_dataset.py:
```diff
@@ -150,7 +149,7 @@ def test_save_load_data(self, gbq_dataset, dummy_dataframe, mocker):
         )
         mocked_read_gbq.assert_called_once_with(
-            sql, project_id=PROJECT, credentials=None
+            project_id=PROJECT, credentials=None, **gbq_dataset._load_args
```
```suggestion
            project_id=PROJECT, credentials=None, query=sql
```

to test that if we don't provide any specific `load_args` that we fallback to the default `query`

mzjp2 commented on tests/extras/datasets/pandas/test_gbq_dataset.py:
```diff
@@ -135,7 +135,6 @@ def test_str_representation(self, gbq_dataset, save_args, load_args):
 
     def test_save_load_data(self, gbq_dataset, dummy_dataframe, mocker):
```
can you add a test where we you change `load_args["query"]` (see e.g above) and make sure that `read_gbq` is called with the right thing?

ajb7 commented:
- Thank you for reviewing :smile:
- I have added parameterized test case for `load_args['query']`
- seems like stringify, try/except, etc. were not needed and already taken care of. Reverted them  😬

mzjp2 commented on tests/extras/datasets/pandas/test_gbq_dataset.py:
```diff
@@ -150,10 +150,25 @@ def test_save_load_data(self, gbq_dataset, dummy_dataframe, mocker):
+        loaded_data = gbq_dataset.load()
+
+        mocked_read_gbq.assert_called_once_with(
+            project_id=PROJECT, credentials=None, query=load_args["query"]
```
I’d check query=“Select 1” here but up to you!

mzjp2 commented:
Thanks again @ajb7

ajb7 commented:
Thank you Kedro Team!!! 😄 :star:
